### PR TITLE
Improve help text formatting

### PR DESCRIPTION
**Pull request by L7R7**
This is a first iteration towards a help text output that works better for parsers with nested commands.

I think what should be improved is the fact that now the braces around the commands and the rest of the mandatory fields are rather hard to parse (it's correct, but not easy to see).
Also, I thought about inserting line breaks between the help and version flags, but it makes things more difficult because we probably have to indent the commands then so it's clear which one's which. The same goes for parsers with deeper nested subcommands. There are braces around the different subcommands now (which helps in distinguishing them), but if the subcommands themselves require several configurations, things get messy again.

### CODE REVIEW

**Comment by L7R7** [comment]
So, for example:

```
sub-commands (-h|--help) | --version | (--config-file FILE_PATH 
   top --name NAME 
 | sub ( a | b ) ( c | d ))
```

could maybe become something like

```
sub-commands
    (-h|--help)
 | --version
 | (--config-file FILE_PATH 
     top --name NAME 
   | sub ( a | b ) ( c | d )
   )
```

(which doesn't look that great, now that I see it)

**Comment by NorfairKing** [comment]
It's not so easy to get this to look nice apparently :P
I'm very open to suggestions but I'm not sure if there's an easy solution.

**Comment by L7R7** [comment]
Yeah, I think it makes sense to start with the change in this PR and then see how it works in real life use cases. I'm happy to do more PRs when I find better ideas.

As a side note: I think optparse-applicative does not have this issue because by default it lists only the names of the top-level commands. But I like the way opt-env-conf does it  where the commands are shown with their options and subcommands because I often end up invoking the command-specific help text only to look up the options for a given command. So I guess there's nothing substantial to learn from optparse-applicative other than maybe "don't list the options at all"

**Comment by NorfairKing** [comment]
Ah yes, here I see how `optparse-applicative` does it:

```
smos-query --help
Usage: smos-query COMMAND [--config-file FILE_PATH] 
                  [--workflow-dir DIRECTORY_PATH] [--archive-dir DIRECTORY_PATH]
                  [--projects-dir DIRECTORY_PATH] 
                  [--archived-projects-dir DIRECTORY_PATHPATH]

  
  Smos Query Tool version: 0.11.0

  Oldest parseable Smos data format version: 0.0.0
  Newest parseable Smos data format version: 2.0.0

Available options:
  -h,--help                Show this help text
  --config-file FILE_PATH  The config file to use
  --workflow-dir DIRECTORY_PATH
                           The workflow directory to use
  --archive-dir DIRECTORY_PATH
                           The archive directory to use
  --projects-dir DIRECTORY_PATH
                           The projects directory to use
  --archived-projects-dir DIRECTORY_PATHPATH
                           The archived projects directory to use

Available commands:
  entry                    Select entries based on a given filter
  report                   Run prepared reports
  waiting                  Print the "WAITING" tasks
  next                     Print the next actions
  ongoing                  Print the ongoing entries
  clock                    Print the clock table
  agenda                   Print the agenda
  projects                 Print the projects overview
  stuck                    Print the stuck projects overview
  work                     Show the work overview
  free                     Find a free slot for a meeting
  log                      Print a log of what has happened.
  stats                    Print the stats actions and warn if a file does not
                           have one.
  tags                     Print all the tags that are in use
```

I guess we could use the same "only one level deep" strategy to make that a bit prettier.

**Comment by L7R7** [comment]
Let me give that a try, I'll see what I can come up with. I might change one of the parsers used in the tests or add a new one with deeper nesting

**Comment by L7R7** [comment]
I pushed new changes so that only the names of the top level commands are shown, including an indentation after the newlines to make it easier to read.

I also tried the changes with the help text of my project, this is how it looks like with the latest commit:

```
Usage: gitlab-helper (-h|--help) 
 | --version 
 | (--config-file FILE_PATH --group-id GROUP_ID --base-url URL --api-token TOKEN --exclude-projects ID1,ID2,ID3 
     show-branches 
   | show-projects 
   | list-all-projects-meta 
   | list-projects-meta 
   | enable-source-branch-deletion 
   | enable-all-discussions-must-be-resolved-for-merge-requirement 
   | enable-successful-pipeline-for-merge-requirement 
   | show-schedules 
   | show-merge-requests 
   | count-deployments 
   | set-merge-method-to-fast-forward 
   | update-merge-requests)
```

Without the changes of today's commit, it looked like that:

```
Usage: gitlab-helper (-h|--help) | --version | (--config-file FILE_PATH --group-id GROUP_ID --base-url URL --api-token TOKEN --exclude-projects ID1,ID2,ID3 
   show-branches 
 | show-projects 
 | list-all-projects-meta 
 | list-projects-meta 
 | enable-source-branch-deletion --execute|-x 
 | enable-all-discussions-must-be-resolved-for-merge-requirement --execute|-x 
 | enable-successful-pipeline-for-merge-requirement --execute|-x 
 | show-schedules 
 | show-merge-requests --[no-]recheck-merge-status 
 | count-deployments --year YEAR --[no-]include-archived-projects 
 | set-merge-method-to-fast-forward --execute|-x 
 | update-merge-requests ( rebase | merge --skip-ci | draft | ready | list ) -u|--user-id USER_ID (--search-title TXT) | (-s|--search TXT) --[no-]recheck-merge-status --execute|-x)
```

Let me know what you think!

**Comment by NorfairKing** [comment]
> Let me know what you think!

It's definitely better than before but I think we can do better still :D
I think we should try following `optparse-applicative`s example a bit closer.
I don't want to tell you what to do though, so let me know what you're up for :)

**Comment by NorfairKing** [comment]
Closed by accident.

**Comment by L7R7** [comment]
> It's definitely better than before but I think we can do better still :D

yes, absolutely!

Do you have something particular in mind to get closer to `optparse-applicative`? Maybe I should try a variant that uses `COMMAND` in the first line and then a compact list of command names below?

In general, I think there's a tradeoff to be made between something that works well on programs with only a few, simple settings and something that works best on programs with lots of complex settings. And I think I would say that it makes more sense to optimize for the complex settings.

**Comment by L7R7** [comment]
Alright, next changes, next questions :smile: 

> I think we can try ignoring the "and" and "or" aspect of the parser for the top-level and then split up the commands (only the first level) from the rest.

That's the idea I followed, and I think it's nice because it is more compact now.

> I'm not sure that works well as long as we haven't done the per-subcommand --help text as well.

If the current approach is the one we're going with, it would be necessary to have that, because otherwise the user has no chance to see what options there are for a subcommand.

There are a couple of things that I would need to to, but I wanted to get your feedback before I go into too much detail:
* I think there should be a similar handling for an empty list of options, similar to the empty list of commands
* There are two empty lines between the settings and the commands. I think I know why (added a todo comment) and I'll probably find a way to fix that
* The indentation of the command list is very weird in cases where the command names are long. We should find a better way than the `\t\t` I used for now

what do you think?

File: opt-env-conf-test/test_resources/docs/hidden/reference.txt [hunk 1]
```diff
@@ -9,6 +9,9 @@
   switch: [37m--version[m
   
 
+[36mAll commands[m:
+no commands available
```
Comment (NorfairKing) [inline opt-env-conf-test/test_resources/docs/hidden/reference.txt:13]: I'm not sure we want this here.
Maybe we want to omit the section instead?

File: opt-env-conf-test/test_resources/docs/sub-commands/man.txt [hunk 2]
```diff
@@ -23,31 +21,10 @@ sub-commands (-h|--help)
...
-    command: c
-    
-    D
-    command: d
-    
-  
+
+.Sh COMMANDS
+  top		command without subcommands
+  sub		command with subcommands
```
Comment (NorfairKing) [inline opt-env-conf-test/test_resources/docs/sub-commands/man.txt:27]: I think we definitely want to see _everything_ in the manpage.
Unless we want mutiple manpages, one per command?

File: opt-env-conf-test/test_resources/docs/sub-commands/reference.txt [hunk 3]
```diff
@@ -15,31 +13,10 @@
...
-    command: [35mc[m
-    
-    [34mD[m
-    command: [35md[m
-    
-  
+
+[36mAll commands[m:
+  [35mtop[m		[34mcommand without subcommands[m
+  [35msub[m		[34mcommand with subcommands[m
```
Comment (NorfairKing) [inline opt-env-conf-test/test_resources/docs/sub-commands/reference.txt:20]: The reference docs should definitely definitely show every thing.

File: opt-env-conf-test/test_resources/docs/yes-no/help.txt [hunk 4]
```diff
@@ -23,3 +23,6 @@ yesNoSwitch example
       [33m<boolean>[m
 
 
+[36mAvailable commands[m:
+no commands available
```
Comment (NorfairKing) [inline opt-env-conf-test/test_resources/docs/yes-no/help.txt:27]: I don't think we want this section in the help text if there are no commands.

File: opt-env-conf/src/OptEnvConf/Doc.hs [hunk 5]
```diff
@@ -321,7 +321,9 @@ renderManPage progname version progDesc docs =
               [".Sh ", "SYNOPSIS"],
               renderShortOptDocs progname optDocs,
               [".Sh ", "SETTINGS"],
-              renderSetDocs docs
+              renderSetDocs docs,
+              [".Sh ", "COMMANDS"],
+              renderCommandDocs docs
```
Comment (NorfairKing) [inline opt-env-conf/src/OptEnvConf/Doc.hs:326]: I like the idea of having a `COMMANDS` section, but:

1. Not if there are no commands
2. They should list the per-command options there

**Review by NorfairKing — COMMENTED** [review]
I really like where this is going!

**Comment by L7R7** [comment]
I just added dynamic column width depending on the command name lengths. It's just a minor cosmetic thing but I thought it's nice to have.

I agree with your comments on the code and will work them in, thanks!

**Comment by L7R7** [comment]
It's impossible to have a parser without any options, right? At least `--version` and `-h` will always be there

**Comment by NorfairKing** [comment]
> It's impossible to have a parser without any options, right? At least `--version` and `-h` will always be there

Technically no. You can run a parser without the `internalParser` being added. But I'm ok with that not being supported very well.

**Comment by L7R7** [comment]
Started to include the full command documentation in the reference and the man page, but it turns out to be a bit tricky to filter out top level settings while keeping the nested ones

File: opt-env-conf/src/OptEnvConf/Doc.hs [hunk 6]
```diff
@@ -469,42 +469,46 @@ renderSetDocs = unlinesChunks . go
...
           Just h ->
             let (sds, rest) = goSameHelp h ds
-            in concat
-                  [ indent $ renderSetDocHeader (Just h),
-                    indent $ concatMap renderSetDocWithoutHeader $ d : sds,
-                    [[]],
-                    goOr rest
+             in concat
+                  [ concat
+                      [ concat
```
Comment (NorfairKing) [inline opt-env-conf/src/OptEnvConf/Doc.hs:503]: Sounds like a beat :p

File: opt-env-conf/src/OptEnvConf/Doc.hs [hunk 7]
```diff
@@ -400,32 +414,32 @@ renderVersionPage progname version =
...
+          ]
+      ]
 
 renderSetDocs :: AnyDocs SetDoc -> [Chunk]
 renderSetDocs = unlinesChunks . go
   where
     go :: AnyDocs SetDoc -> [[Chunk]]
     go = \case
-      AnyDocsCommands cs -> concatMap goCommand cs
+      AnyDocsCommands _ -> [] -- todo: this empty list causes the double newline between settings and commands
```
Comment (L7R7) [inline opt-env-conf/src/OptEnvConf/Doc.hs:438]: This is a part that annoys me a bit, this is the line that causes the double newline between the settings and the commands block but so far I failed to find a way to avoid it without removing other empty lines

File: opt-env-conf-test/test_resources/docs/sub-commands/man.txt (hunk 2 above)
Comment (L7R7) [reply opt-env-conf-test/test_resources/docs/sub-commands/man.txt]: I think a single manpage is the more idiomatic way here? I don't use manpages that often but I think it's not unusual for manpages being exhaustive and therefore longer

File: opt-env-conf/src/OptEnvConf/Doc.hs (hunk 7 above)
Comment (NorfairKing) [reply opt-env-conf/src/OptEnvConf/Doc.hs:438]: @L7R7 This is a very common problem when producing output.
The "usual" way of dealing with it is to wrap the result in another layer of `Maybe` or `[]` so you can remove that layer before rendering.

Comment (L7R7) [reply opt-env-conf/src/OptEnvConf/Doc.hs:438]: I tried that, but I failed. I was sure several times that I have the right solution but the tests disagreed :smile:

**Comment by L7R7** [comment]
I think this is coming close to a "feature-complete" version. I also think that this change is closely related to the per-command help output (do you already have a plan on how this would work?)

**Comment by NorfairKing** [comment]
> I think this is coming close to a "feature-complete" version. I also think that this change is closely related to the per-command help output (do you already have a plan on how this would work?)

I have some ideas but no time atm.
The idea is that the help parsing would happen at multiple levels of the parser and the parser would be able to return early with a "help invoked here, at this command level" constructor.

**Comment by L7R7** [comment]
> The idea is that the help parsing would happen at multiple levels of the parser and the parser would be able to return early with a "help invoked here, at this command level" constructor.

that sounds good!

> I have some ideas but no time atm.

Theoretically I could spend some more time working on this library, but on the one hand I'll be away from computers for at least the next week (maybe longer) and on the other hand I don't know if it makes sense to work on that feature because I'll probably need some guidance from you

**Comment by NorfairKing** [comment]
> > I have some ideas but no time atm.
> 
> Theoretically I could spend some more time working on this library, but on the one hand I'll be away from computers for at least the next week (maybe longer) and on the other hand I don't know if it makes sense to work on that feature because I'll probably need some guidance from you

I appreciate it!
Let me try some things, try to merge this, and get back to you, ok?

**Comment by L7R7** [comment]
> Let me try some things, try to merge this, and get back to you, ok?

Yes, works for me! If you happen to find something to make progress, go ahead. Don't wait for me, I'm happy with any progress regardless of whether I am contributing or not

**Comment by NorfairKing** [comment]
@L7R7 I think I figured it out.
I'll make a PR so you can have a look and rebase onto it.

**Comment by NorfairKing** [comment]
Merged manually.
